tiledb/api/src/query/subarray.rs

```diff
@@ -99,18 +99,23 @@ impl<'ctx> Builder<'ctx> {
                 })
             }
         }?;
-        self.build()
+        Ok(self)
     }
 
-    fn build(mut self) -> TileDBResult<QueryBuilder<'ctx>> {
+    /// Apply the subarray to the query, returning the query builder.
+    pub fn to_query(self) -> TileDBResult<QueryBuilder<'ctx>> {
         let c_context = self.subarray.context.capi();
         let c_query = *self.query.query.raw;
         let c_subarray = *self.subarray.raw;
 
         self.capi_return(unsafe {
             ffi::tiledb_query_set_subarray_t(c_context, c_query, c_subarray)
         })?;
-        self.query.query.subarrays.push(self.subarray);
         Ok(self.query)
     }
+
+    /// Apply the subarray to the query and finish constructing the query.
+    pub fn build(self) -> TileDBResult<Query<'ctx>> {
```

Comment:
Wouldn't this force users to have to know that the subarray being set must go last on the query builder? That seems like a not very awesome requirement. I've stumbled on this "sub-builder" pattern a couple of times and generally end up avoiding it by using other means. Either by reshaping the "parent" builder's API or extracting things to accept a "pre-built" value.

I.e., I'd prefer something like either of these:

```rust

query.do_stuff()?
    .subarray(&[
        ("rows", &[1i32, 2]),
        ("columns" &[2i32, 4])
    ])
    .do_more_stuff()?;
    .build()?;
```

Or perhaps:

```rust
let subarray = SubarrayBuilder::new(&ctx)
    .dimension::<i32>("rows", 1, 2)?
    .dimension::<i32>("columns", 2, 4)?
    .build();

let query = builder.do_stuff()?
    .subarray(subarray)
    .do_other_stuff()?
    .build();
````

On the other hand, if the subarray forces our hand, I could also see having a different API for subarray as well.

```rust
let query = builder.do_stuff()?
    .start_subarray()
    .dimension::<i32>("rows", 1, 2)?
    .dimension::<i32>("columns", 2, 4))?
    .finish_subarray()?
    .do_other_stuff()?
    .build();

Reply:
> Wouldn't this force users to have to know that the subarray being set must go last on the query builder?

This is not required as `to_query` can be used to get the query builder back.  I just made `build` do this as a convenience, but if it is a confusing convenience then I can revert it.  Where you suggest `start_subarray` and `finish_subarray`, what's in the PR is `add_subarray` and `to_query` - I like your nomenclature better.